### ISSUE TITLE
Windows + non ansi symbols in filename path: not possible to open fits file

### ISSUE THREAD

**Issue by art-den** [simonrw/rust-fitsio#190]
`Fits error: FitsError { status: 104, message: "could not open the named file" }`
Any ideas how to fix it?

**Comment by simonrw** [comment]
Hi, thanks for the report. Do you have an example filename?

Because we use a c library to do the heavy lifting, we may struggle. I'm not at my computer right now but I'll check when I can.

There might be a way to open a file from a.file descriptor that we can use.

**Comment by cjordan** [comment]
I thought the issue might've been because the specified filename isn't valid UTF-8, and (to my knowledge) the only way to convert an `AsRef<Path>` to a `CString` is with a valid UTF-8 filename.

https://github.com/simonrw/rust-fitsio/blob/1a2c4589ac223581a89663bfb0c15ba99b56cd77/fitsio/src/fitsfile.rs#L52-L59

... but that would've triggered the `except` in that code block.  @art-den can you give the file name you tried to use?  These error messages are pretty poor but the issue may be as simple as what's said on the tin.

**Comment by art-den** [comment]
Examples:
`D:\test\Путь\HA_dzX2_KS_D_eco.fits`
`D:\test\Path\Имя_файла.fits`

**Comment by simonrw** [comment]
So it's not the non-ascii characters in the filenames - I've tested with `Имя_файла.fits` and it opens just fine in `cfitsio` and this crate. I suspect it's something to do with windows absolute paths, but I have recently deleted my windows VM so it might be a while before I can test.

@art-den can you try using relative paths (e.g. `..\Имя_файла.fits`) to see if that still works?

**Comment by simonrw** [comment]
Right: my windows VM still exists, and it cfitsio (the c library) struggles to open `Имя_файла.fits` after all. `cfitsio` expects the filename to be a `char*` (equivalent to rust `u8`) and does not seem to be doing any sort of unicode handling. In fact there's no mention of the word "unicode" in the `cfitsio` source code at all.

Windows unicode is utf-16 which is not what Rust or C expect at all. I will bow to someone with more windows experience than me but a workaround is to save files with ascii filenames for now.

**Comment by art-den** [comment]
Astrophotography software saves fits files into user's folders like `C:\Users\<user name>\Desktop\SharpCap Captures` (or something else) by default. And <user name> can contains not only ascii characters ((
I will try to set default folder to `C:\Users\<user name>\Desktop\SharpCap Captures` and to open fits file by short file name (without full path). It will not solve problem with non-ascii in short file name but I think this is very rare situation.

**Comment by simonrw** [comment]
I can't immediately think of a way around this. I recognise that the software has no control over the username of the user, but I think there would be a problem with `cfitsio` as well so we're not _adding_ any problems here.

It's not ideal, but I don't understand enough about windows paths and non-ascii characters. Is this something that you can live with? Also: could you map the  path to e.g. `Z:\` so it does not contain the username?

**Comment by art-den** [comment]
My heavyweight solution to the problem ))
```rust
#[cfg(target_os = "windows")]
mod win_fits_open {
    use std::{sync::Mutex, path::PathBuf};

    pub static OPEN_FITS_MUTEX: Mutex<()> = Mutex::new(());

    pub struct CurDirRestore {
        cur_dir: PathBuf,
    }

    impl CurDirRestore {
        pub fn new() -> anyhow::Result<Self> {
            Ok(Self {
                cur_dir: std::env::current_dir()?,
            })
        }
    }

    impl Drop for CurDirRestore {
        fn drop(&mut self) {
            _ = std::env::set_current_dir(&self.cur_dir);
        }
    }
}

#[cfg(target_os = "windows")]
fn fits_file_open_helper<F>(file_name: &Path, fun: F) -> anyhow::Result<FitsFile>
where F: Fn(&Path) -> anyhow::Result<FitsFile> {
    use win_fits_open::*;
    let result = match (file_name.parent(), file_name.file_name()) {
        (Some(file_dir), Some(rel_name)) => {
            let change_dir_locker = OPEN_FITS_MUTEX.lock();
            let cur_dir_restore = CurDirRestore::new()?;
            std::env::set_current_dir(file_dir)?;
            let rel_name = rel_name.to_str().unwrap_or("");
            let fits = fun(&Path::new(rel_name))?;
            drop(cur_dir_restore);
            drop(change_dir_locker);
            fits
        },
        _ =>
            fun(file_name)?,
    };
    Ok(result)
}

#[cfg(not(target_os = "windows"))]
fn fits_file_open_helper<F>(file_name: &Path, fun: F) -> anyhow::Result<FitsFile>
where F: Fn(&Path) -> anyhow::Result<FitsFile> {
    fun(file_name)
}

//////////////////////////////////////////////////

// using this

let mut fptr = fits_file_open_helper(
    file_name,
    |file_name| Ok(FitsFile::open(file_name)?)
)?;

// or

let mut fptr = fits_file_open_helper(
    file_name,
    |file_name| {
        Ok(FitsFile::create(file_name)
            .with_custom_primary(&image_description)
            .open()?)
    }
)?;
```

**Comment by simonrw** [comment]
I'm sorry you have to work around this in this way, but that seems reasonable since `cfitsio` doesn't support non-ascii filenames :facepalm: 

In general - are you happy that this fixes your problem going forwards? If so I will close this issue (and it will remain forever as documentation for others), otherwise I'll leave this open and we can consider a "proper" fix

**Comment by art-den** [comment]
It looks like this is issue of msys2 (and libfitsio also). My workaround fixes most common problem. So i'm happy ))
